src/tests/test_dal_middleware.py

```diff
@@ -0,0 +1,303 @@
+    # needs to be run 'last'.
+
+    class FooMiddleware(MethodMiddleware):
+        pass
+
+    class BarMiddleware(MethodMiddleware):
+        pass
+
+    class BazMiddleware(MethodMiddleware):
+        pass
+
+    middlewares = [FooMiddleware, BarMiddleware, BazMiddleware]
+    dal = DataAccessLayer(data_manager=None, middleware=list(middlewares))
+
+    # The DAL handles making sure all of these are instantiated, but
+    # we have to manually do it for testing here.
+    middlewares.append(DalMethodResolverMiddleware)
+    expected_middlewares = []
+    for m in reversed(middlewares):
+        expected_middlewares.append(m())
+
+    # Verify if the objects in dal._reversed_middleware are instances of the
+    # expected middleware types in the right order
+    for i in range(0, len(expected_middlewares)):
+        assert isinstance(dal._reversed_middleware[0], type(expected_middlewares[0]))
+
+
+def test_dal_method_middleware():
+    """
+    Verify that when calling a method on a service off the DAL,
+    the method middlewares run, in the correct order and run
+    before the method itself is called.
+
+
+    We deliberately do not use the DalMethodResolverMiddleware here,
+    because we are testing middleware call chains, not functionality
+    of that class.
+    """
+    middleware_requests = []
+
+    def test_method():
+        middleware_requests.append(('default-handler-call', None))
+
+    class OuterMiddleware(MethodMiddleware):
+        def __call__(self, request, handler):
+            middleware_requests.append(('outer-middleware-ingress', request))
+            result = handler(request)
+            middleware_requests.append(('outer-middleware-egress', request))
+            return result
+
+    class SecondMiddleware(MethodMiddleware):
+        def __call__(self, request, handler):
+            middleware_requests.append(('second-middleware-ingress', request))
+            result = handler(request)
+            middleware_requests.append(('second-middleware-egress', request))
+            return result
+
+    # Verify custom default middleware
+    class DefaultMiddleware(MethodMiddleware):
+        def __call__(self, request, handler):
+            middleware_requests.append(('default-middleware-ingress', request))
+            # We need to get past the DalMethodResolverMiddleware
+            # which sets the `dal_method` attribute to
+            # the method that has to be called
+            request.dal_method = test_method
+            result = handler(request)
+            middleware_requests.append(('default-middleware-egress', request))
+            return result
+
+    dm = DataManager()
+
+    dal = DataAccessLayer(data_manager=dm, middleware=[OuterMiddleware, SecondMiddleware],
+                          default_middleware=(DefaultMiddleware,))
+
+    with dm.context():
+        dal.fake.service.method()
+
+    first_middleware_to_run = middleware_requests[0]
```

Comment:
`request` is never used. If you drop it from `middleware_requests` , verification becomes as simple as:

```
assert middleware_requests == [
'outer-middleware-ingress',
'second-middleware-ingress',
'default-middleware-ingress',
# etc
]
```

Reply:
one of the other things I said yesterday, "I'm sure Mike will be able to simplify some of this test stuff, he always makes stuff so clear in his tests."